### DISCUSSION TITLE
About Digest Authentication

### DISCUSSION THREAD

**Discussion by JayDommaschk** [onvif/testspecs#5] [Q&A] [answered]
Hello,

I am having a question regarding digest authentication. I originally asked this in a ticket I sent via Onvif's contact form, but in their reply they told me to post my question here instead. So here we go:

I am implementing Onvif Profile T for devices. While doing the digest authentication part, I realised that the Onvif Core Specification only mentions RFC2617 but not RFC7617. That would imply that I can only use MD5 hashes for storing passwords, not SHA-256 hashes. Since MD5 is known to be insecure and my company places highest priority on data protection and security, I would like to know if there will be any way for us to get the official Onvif certification if we only support SHA-256.

Thanks for any reply. Have a nice day.

**Comment by baldvin137** [comment, marked as answer]
Good question. 

ONVIF does not specify how devices shall work internally, ONVIF only specifies the communication between devices and clients. Although MD5 is cryptographically broken, it is still widely used for digest authentication while there is to my knowledge little support for SHA-256 digest authentication on the client side. So unfortunately using SHA-256 only for digest authentication would not provide interoperability and therefore MD5 is required for Profile T. As noted in the ONVIF Core Specification (section 5.9.1 Authentication): _"Both digest authentication and the user name token profile give only a rudimentary level of security. In a system where security is important, it is recommended to always configure the device for TLS-based access (see Advanced Security Service)."_

So my answer would be that you have to support MD5 for digest authentication to provide interoperability for Profile T clients, and if you want higher security you should implement TLS. 

Note also that Profile T does not specify default out-of-the-box behavior, nor does it exclude any optional behavior, such as the use of SHA-256. A device could have MD5 disabled by default and by default only use SHA-256 and still be Profile T compliant, as long as the device can be put in a Profile T compliant state. How to make the device Profile T compliant would then have to be described in the mandatory ONVIF Interface Guide.

Hope this helps!

**Reply by JayDommaschk** [reply]
Hi,

that helps indeed. Thanks a lot.

The main lesson for me is that I will not get around implementing MD5. I will certainly give the user the option to disable MD5, but having it disabled by default would seem to go against the basic idea of onvif, that is, that devices should work out of the box without need for any vendor specific interface. 

TLS will also be implemented, but as far as I know that only secures communication. My concern is with the case that the device gets somehow compromised so that attackers can read out the (hashed) passwords. Not that that would be a likely case, but bad things happen all the time, better be safe than sorry. (then again, I guess the point can be made that if a device is compromised to that point, there are much bigger issues and attackers might not even need passwords anymore)

This is just some rambling by me, though. I understand that onvif is the way it is and existing profiles cannot be just changed to account for concerns that came up after a profile is finalized, not in a non-backwards compatible way anyway.

So thanks a lot for taking the time to answer me. Have a nice day.

**Reply by baldvin137** [reply]
Happy to help!

Yes, I agree that if a device is compromized, you probably have bigger problems than the password hashing. Typically ONVIF devices should not have a big database of users and passwords. In fact ONVIF systems should (in my opinion) be setup so that devices have unique usernames and passwords so that if a device gets compromized nothing is lost except that device. But that kind of system design aspects is out of scope of the current ONVIF specifications.
